vms/platformvm/state/subnet_only_validator.go

```diff
@@ -128,3 +176,119 @@ func putSubnetOnlyValidator(db database.KeyValueWriter, vdr SubnetOnlyValidator)
 func deleteSubnetOnlyValidator(db database.KeyValueDeleter, validationID ids.ID) error {
 	return db.Delete(validationID[:])
 }
+
+type subnetOnlyValidatorsDiff struct {
+	numAddedActive      int               // May be negative
+	modifiedTotalWeight map[ids.ID]uint64 // subnetID -> totalWeight
+	modified            map[ids.ID]SubnetOnlyValidator
+	modifiedHasNodeIDs  map[subnetIDNodeID]bool
+	active              *btree.BTreeG[SubnetOnlyValidator]
+}
+
+func newSubnetOnlyValidatorsDiff() *subnetOnlyValidatorsDiff {
+	return &subnetOnlyValidatorsDiff{
+		modifiedTotalWeight: make(map[ids.ID]uint64),
+		modified:            make(map[ids.ID]SubnetOnlyValidator),
+		modifiedHasNodeIDs:  make(map[subnetIDNodeID]bool),
+		active:              btree.NewG(defaultTreeDegree, SubnetOnlyValidator.Less),
+	}
+}
+
+// getActiveSubnetOnlyValidatorsIterator takes in the parent iterator, removes
+// all modified validators, and then adds all modified active validators.
+func (d *subnetOnlyValidatorsDiff) getActiveSubnetOnlyValidatorsIterator(parentIterator iterator.Iterator[SubnetOnlyValidator]) iterator.Iterator[SubnetOnlyValidator] {
+	return iterator.Merge(
+		SubnetOnlyValidator.Less,
+		iterator.Filter(parentIterator, func(sov SubnetOnlyValidator) bool {
+			_, ok := d.modified[sov.ValidationID]
+			return ok
+		}),
+		iterator.FromTree(d.active),
+	)
+}
+
+func (d *subnetOnlyValidatorsDiff) hasSubnetOnlyValidator(subnetID ids.ID, nodeID ids.NodeID) (bool, bool) {
+	subnetIDNodeID := subnetIDNodeID{
+		subnetID: subnetID,
+		nodeID:   nodeID,
+	}
+	has, modified := d.modifiedHasNodeIDs[subnetIDNodeID]
+	return has, modified
+}
+
+func (d *subnetOnlyValidatorsDiff) putSubnetOnlyValidator(state Chain, sov SubnetOnlyValidator) error {
+	var (
+		prevWeight uint64
+		prevActive bool
+		newActive  = sov.isActive()
+	)
+	switch priorSOV, err := state.GetSubnetOnlyValidator(sov.ValidationID); err {
+	case nil:
+		if !priorSOV.constantsAreUnmodified(sov) {
+			return ErrMutatedSubnetOnlyValidator
+		}
+
+		prevWeight = priorSOV.Weight
+		prevActive = priorSOV.isActive()
+	case database.ErrNotFound:
+		// Verify that there is not a legacy subnet validator with the same
+		// subnetID+nodeID as this L1 validator.
+		_, err := state.GetCurrentValidator(sov.SubnetID, sov.NodeID)
+		if err == nil {
+			return ErrConflictingSubnetOnlyValidator
+		}
+		if err != database.ErrNotFound {
+			return err
+		}
+
+		has, err := state.HasSubnetOnlyValidator(sov.SubnetID, sov.NodeID)
+		if err != nil {
+			return err
+		}
```

Comment:
are we handling the `database.NotFound` err in the `HasSubnetOnlyValidator` (and returning false)?

Reply:
This aligns with `Get` and `Has` behave differently on the database interface. `Get` returns `ErrNotFound` if the entry is missing. `Has` explicitly returns `true` if the entry exists and `false` if the entry does not exist.

`Has` never returns `ErrNotFound`.